### PR TITLE
Add per-epoch withdraw caps

### LINKED ISSUE #3: Vault: Add per-epoch withdraw caps for each vault
Add per-epoch withdraw caps on a vault basis as a pro-rata share of the vault stake which prevents excess withdraws

### PR DESCRIPTION
**Pull request by aoikurokawa**
#### Motivation
- Closes #3 

#### Solution
- Add `epoch_withdraw_cap`, `current_epoch`, `epoch_withdrawn_amount` in `Vault` struct.
- Does not allow withdrawing over `epoch_withdraw_cap`

### CODE REVIEW

**Comment by aoikurokawa** [comment]
@buffalu 
Can you review this PR?
Thank you so much!

**Comment by coachchucksol** [comment]
Hey @aoikurokawa! Thanks for the contribution!

We've had some discussions around this and need the solution to #3 to be as follows:

The `vault` struct needs the following: `epoch_withdraw_cap_bps`, `last_slot_updated`, `epoch_withdraw_amount`, `epoch_snapshot_amount` along with the function `is_update_needed`

- `epoch_withdraw_cap_bps` - the percentage 25% - 100% (2500 - 10000) that is the max that can be withdrawn from the vault based on a snapshot of assets in the vault at the beginning of an epoch
- `last_slot_updated` - The slot when `process_update_vault` is called at the beginning of an epoch
- `epoch_withdraw_amount` - The tally of assets withdrawn on that epoch, this cannot be above `epoch_snapshot_amount` x  `epoch_withdraw_cap_bps`
- `epoch_snapshot_amount` - The amount of assets in the vault at the time of calling `process_update_vault`
- `is_update_needed` - Check that `Clock.slots` / `config.epoch_length` ( See`is_update_needed` in VaultDelegationList )

Flow:

At the beginning of the epoch ( When  `vault.is_update_needed` is true ) we run `process_update_vault` this will:
- Set `last_slot_updated` to `Clock.slot`
- Clear `epoch_withdraw_amount` to 0
- Set `epoch_snapshot_amount` to whatever is available to withdraw from the vault

When a burn or enqueue withdrawal ticket happens:
- Check that `vault.is_update_needed` is false
- Check that `epoch_withdraw_amount` + `amount_to_withdraw` <= `epoch_snapshot_amount` * `epoch_withdraw_cap_bps`
- Increment `epoch_withdraw_amount` by `amount_to_withdraw`

Hope this all makes sense!

**Comment by aoikurokawa** [comment]
@buffalu @coachchucksol 
Thank you for explanation!
Ready for review it! Thank you!

**Comment by coachchucksol** [comment]
Hi! I'll take a look at it today!

**Review by coachchucksol — CHANGES_REQUESTED** [review]
Great work so far! Just a couple things to change - also, please make sure to resolve any merge issues :)

File: integration_tests/tests/vault/burn_withdrawal_ticket.rs
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply integration_tests/tests/vault/burn_withdrawal_ticket.rs:1]: Thank you! I will keep all tests except `epoch_withdraw_cap_bps` part.

File: integration_tests/tests/vault/set_admin.rs [hunk 1]
```diff
@@ -18,6 +18,7 @@ mod tests {
         let deposit_fee_bps = 99;
         let withdrawal_fee_bps = 100;
         let reward_fee_bps = 101;
+        let epoch_withdraw_cap_bps = 101;
```
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply integration_tests/tests/vault/set_admin.rs:21]: Sorry, I changed to `let epoch_withdraw_cap_bps = 102;`

File: vault_core/src/vault.rs [hunk 2]
```diff
@@ -539,6 +572,36 @@ impl Vault {
...
+    /// * `amount_to_withdraw` - A `u64` representing the amount of tokens that is requested to be withdrawn.
+    ///
+    /// # Returns
+    /// * `Result<(), VaultError>` - Returns `Ok(())` if the withdrawal is within the allowed limit.
+    ///
+    /// # Errors
+    /// This function can return the following errors:
+    /// * [`jito_vault_sdk::error::VaultError::VaultOverflow`] - If any arithmetic operation (addition or multiplication) results in an overflow.
+    /// * [`jito_vault_sdk::error::VaultError::VaultWithdrawalLimitExceeded`] - If the requested withdrawal exceeds the allowed limit for the epoch.
+    pub fn check_withdrawal_allowd(&self, amount_to_withdraw: u64) -> Result<(), VaultError> {
```
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply vault_core/src/vault.rs:587]: I am too :smile_cat: 
I installed https://github.com/crate-ci/typos

File: vault_core/src/vault.rs [hunk 3]
```diff
@@ -1517,10 +1588,35 @@ mod tests {
...
+            0,
         );
 
         let fee = vault.calculate_rewards_fee(1000).unwrap();
 
         assert_eq!(fee, 1000);
     }
+
+    #[test]
+    fn test_check_withdrawal_allowd() {
```
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply vault_core/src/vault.rs:1600]: Thanks. I fixed.

File: vault_core/src/vault.rs [hunk 4]
```diff
@@ -233,6 +246,26 @@ impl Vault {
         self.withdrawal_fee_bps = PodU16::from(withdrawal_fee_bps);
     }
 
+    pub fn epoch_withdraw_amount(&self) -> u64 {
+        self.epoch_withdraw_amount.into()
+    }
+
+    pub fn set_epoch_withdraw_amount(&mut self, epoch_withdraw_amount: u64) {
```
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply vault_core/src/vault.rs:253]: I divided into `clear_epoch_withdraw_amount` and `increment_epoch_withdraw_amount`.

File: vault_program/src/enqueue_withdrawal.rs [hunk 5]
```diff
@@ -109,7 +113,12 @@ pub fn process_enqueue_withdrawal(
     );
 
     vault.increment_vrt_enqueued_for_cooldown_amount(vrt_amount)?;
-
+    vault.set_epoch_withdraw_amount(
```
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply vault_program/src/enqueue_withdrawal.rs:116]: I changed to `increment_epoch_withdraw_amount`.

**Comment by aoikurokawa** [comment]
@coachchucksol 
Thank you for reviewing!!

File: vault_core/src/vault.rs [hunk 6]
```diff
@@ -539,6 +606,48 @@ impl Vault {
...
+        } else {
+            let total_withdraw_amount = self
+                .epoch_withdraw_amount()
+                .checked_add(amount_to_withdraw)
+                .ok_or(VaultError::VaultOverflow)?;
+            let max_allowed_withdraw = self
+                .epoch_snapshot_amount()
+                .checked_mul(self.epoch_withdraw_cap_bps() as u64)
+                .ok_or(VaultError::VaultOverflow)?
+                .div_ceil(10_000);
```
Comment (buffalu) [inline vault_core/src/vault.rs:640]: nit: use constant for this now

File: vault_program/src/initialize_vault.rs [hunk 7]
```diff
@@ -112,6 +113,7 @@ pub fn process_initialize_vault(
             deposit_fee_bps,
             withdrawal_fee_bps,
             reward_fee_bps,
+            epoch_withdraw_cap_bps,
```
[the remark this answers is not in the document]
Comment (buffalu) [reply vault_program/src/initialize_vault.rs:116]: can you add test to try to init with 10001 too?

File: vault_program/src/close_update_state_tracker.rs [hunk 8]
```diff
@@ -68,6 +68,13 @@ pub fn process_close_vault_update_state_tracker(
         vault.increment_vrt_ready_to_claim_amount(vault.vrt_cooling_down_amount())?;
         vault.set_vrt_cooling_down_amount(vault.vrt_enqueued_for_cooldown_amount());
         vault.set_vrt_enqueued_for_cooldown_amount(0);
+        vault.set_epoch_withdraw_amount(0);
+
+        let max_withdrawable = vault
+            .tokens_deposited()
+            .checked_sub(vault.delegation_state.total_security()?)
```
[the remark this answers is not in the document]
Comment (buffalu) [reply vault_program/src/close_update_state_tracker.rs:75]: hmm good question, i think can probably just do tokens deposited here

File: vault_program/src/enqueue_withdrawal.rs [hunk 9]
```diff
@@ -78,6 +78,9 @@ pub fn process_enqueue_withdrawal(
     vault.check_mint_burn_admin(optional_accounts.first())?;
     vault.check_update_state_ok(Clock::get()?.slot, config.epoch_length())?;
 
+    let amount_to_withdraw = vault.calculate_assets_returned_amount(vrt_amount)?;
```
Comment (buffalu) [inline vault_program/src/enqueue_withdrawal.rs:81]: missing a check in the burn instruction

File: vault_core/src/vault.rs [hunk 10]
```diff
@@ -125,12 +125,21 @@ pub struct Vault {
...
+    epoch_snapshot_amount: PodU64,
+
     /// The deposit fee in basis points
     deposit_fee_bps: PodU16,
 
     /// The withdrawal fee in basis points
     withdrawal_fee_bps: PodU16,
 
+    /// The percentage 25% - 100% (2500 - 10000) that is the max that can be withdrawn from the vault based on a snapshot of assets in the vault at the beginning of an epoch
+    pub epoch_withdraw_cap_bps: PodU16,
```
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply vault_core/src/vault.rs:141]: I changed not to expose `epoch_withdraw_cap_bps` directly.

File: vault_program/src/initialize_vault.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply vault_program/src/initialize_vault.rs:116]: I added a new test `test_initialize_vault_with_epoch_withdraw_cap_bps` in `initialize_vault`

File: vault_program/src/enqueue_withdrawal.rs (hunk 9 above)
Comment (aoikurokawa) [reply vault_program/src/enqueue_withdrawal.rs:81]: added the check in burn ix too.
```rust
    let BurnSummary {
        fee_amount,
        burn_amount,
        out_amount,
    } = vault.burn_with_fee(amount_in, min_amount_out)?;

    vault.check_withdrawal_allowed(out_amount)?;

```

File: vault_core/src/vault.rs (hunk 6 above)
Comment (aoikurokawa) [reply vault_core/src/vault.rs:640]: Other functions use `MAX_FEE_BPS`, but in this case, this constant does not match, I think.
So I created new constant `MAX_EPOCH_WITHDRAW_BPS`, though both values are same.

File: vault_core/src/vault.rs [hunk 11]
```diff
@@ -125,12 +125,21 @@ pub struct Vault {
     /// The slot of the last time the delegations were updated
     last_full_state_update_slot: PodU64,
 
+    /// The tally of assets withdrawn on that epoch, this cannot be above epoch_snapshot_amount x epoch_withdraw_cap_bps
+    epoch_withdraw_amount: PodU64,
```
[the remark this answers is not in the document]
Comment (aoikurokawa) [reply vault_core/src/vault.rs:129]: It is supported_token.
I changed the name to `epoch_withdraw_supported_token_amount`

File: vault_core/src/vault.rs [hunk 12]
```diff
@@ -528,6 +603,46 @@ impl Vault {
...
+    /// This function can return the following errors:
+    /// * [`jito_vault_sdk::error::VaultError::VaultOverflow`] - If any arithmetic operation (addition or multiplication) results in an overflow.
+    /// * [`jito_vault_sdk::error::VaultError::VaultWithdrawalLimitExceeded`] - If the requested withdrawal exceeds the allowed limit for the epoch.
+    pub fn check_withdrawal_allowed(&self, amount_to_withdraw: u64) -> Result<(), VaultError> {
+        let epoch_withdraw_amount: u128 = self.epoch_withdraw_supported_token_amount().into();
+        let epoch_snapshot_amount: u128 = self.epoch_snapshot_supported_token_amount().into();
+        let last_full_state_update_slot: u128 = self.last_full_state_update_slot().into();
+        let max_withdrawable = self.tokens_deposited();
+
+        if epoch_snapshot_amount == 0
```
Comment (coachchucksol) [inline vault_core/src/vault.rs:624]: Why have this check? If nothing is in the vault, or nothing has been updated I don't think they should be able to withdraw anything

**Review by coachchucksol — CHANGES_REQUESTED** [review]
If I could get clarity on my question in the function `check_withdrawal_allowed` then I think everything here is good to go!

File: vault_core/src/vault.rs (hunk 12 above)
Comment (aoikurokawa) [reply vault_core/src/vault.rs]: Sorry, I was quite stupid...
You are totally right. I removed the check.

**Pull request closed without merging**